ns for fx: add fp16 function shadowing

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56927 ns for fx: add missing add_relu and mul_relu patterns
* #56762 ns for fx: additional bugfix for user defined functions
* #56742 ns for fx: allow comparing int8 to int8 for functionals
* #56493 ns for fx: add option to skip matching classes and functions
* #56408 ns for fx: support binary ops when adding unshadowed loggers for inputs
* #56384 ns for fx: bug fix for shadowing fp16 emulation patterns
* **#56311 ns for fx: add fp16 function shadowing**
* #56301 ns for fx: allow user functions in shadowing
* #56296 ns for fx: move node I/O dtype mapping to be local instead of global
* #56292 ns for fx: handling for user functions in weight and unshadowed act APIs
* #56283 ns for fx: enable user defined functions for graph matching

Summary:

Adds functionality for shadowing user functions with fp16 I/O dtype.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_user_defined_function
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27836064](https://our.internmc.facebook.com/intern/diff/D27836064)

vkuzo committed:
ns for fx: add fp16 function shadowing

Summary:

Adds functionality for shadowing user functions with fp16 I/O dtype.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_user_defined_function
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
rebase on "ns for fx: add fp16 function shadowing"


Summary:

Adds functionality for shadowing user functions with fp16 I/O dtype.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_user_defined_function
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27836064](https://our.internmc.facebook.com/intern/diff/D27836064)

[ghstack-poisoned]

jerryzh168 commented on test/quantization/test_numeric_suite_fx.py:
```diff
@@ -1338,7 +1345,7 @@ def forward(self, x):
 
         node_type_to_io_type_map = get_node_type_to_io_type_map()
-        node_type_to_io_type_map['funs_io_type_fp32'].add(_wrapped_hardswish)
+        node_type_to_io_type_map['funs_io_type_fp32'].add(_wrapped_hardswish_fp16)
```
_wrapped_hardswich_fp16 seem to have fp16 input and output type, why are we adding this to funs_io_type_fp32?

vkuzo replied:
fixed

vkuzo replied:
sorry, fix upcoming, but ghstack is having some issues, please ignore previous comment

vkuzo committed:
Update on "ns for fx: add fp16 function shadowing"


Summary:

Adds functionality for shadowing user functions with fp16 I/O dtype.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_user_defined_function
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27836064](https://our.internmc.facebook.com/intern/diff/D27836064)

[ghstack-poisoned]

vkuzo committed:
Update on "ns for fx: add fp16 function shadowing"


Summary:

Adds functionality for shadowing user functions with fp16 I/O dtype.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_user_defined_function
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27836064](https://our.internmc.facebook.com/intern/diff/D27836064)

[ghstack-poisoned]

vkuzo committed:
Update on "ns for fx: add fp16 function shadowing"


Summary:

Adds functionality for shadowing user functions with fp16 I/O dtype.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_user_defined_function
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27836064](https://our.internmc.facebook.com/intern/diff/D27836064)

[ghstack-poisoned]
